### PR TITLE
Eliminate CDHashMap::insertAtContextLevelZero

### LINKED ISSUE #4767: "Fatal failure" when overwriting symbols with `:global-declarations true`
## Issue

For the following file:

```
(set-option :global-declarations true)
(push 1)
(declare-fun x!0 () Bool)
(assert x!0)
(check-sat)
(get-model)
(pop 1)
(push 1)
(declare-fun x!0 () (_ BitVec 32))
(assert (= x!0 (_ bv0 32)))
(check-sat)
(get-model)
```

CVC4 falls over as follows:

```
[avj@tempvm build]$ ./bin/cvc4 -m -i ../../../blah.smt2
../../../blah.smt2:2.5: No set-logic command was given before this point.
../../../blah.smt2:2.5: CVC4 will make all theories available.
../../../blah.smt2:2.5: Consider setting a stricter logic for (likely) better performance.
../../../blah.smt2:2.5: To suppress this warning in the future use (set-logic ALL).
sat
(model
(define-fun x!0 () Bool true)
)
Fatal failure within void CVC4::context::CDHashMap<Key, Data, HashFcn>::insertAtContextLevelZero(const Key&, const Data&) [with Key = std::__cxx11::basic_string<char>; Data = CVC4::Expr; HashFcn = std::hash<std::__cxx11::basic_string<char> >] at ../src/./context/cdhashmap.h:388
Check failure

 d_map.find(k) == d_map.end()

Aborted (core dumped)
```

`git rev-parse --short HEAD` == e8df6f67c

### PR DESCRIPTION
**Pull request by ajreynol**
This method unnecessarily complicates the usage of CDHashMap. It had one usage in the current code, for dealing with global declarations. However, this was an entirely artificial use case, as one should properly manage scopes when this option is true (i.e. global-declarations simply disables user-level scoping in the symbol table).

It also simplifies the symbol table so that it doesn't automatically push a global outermost scope. Instead, this scope is pushed when the logic is declared, so that background symbols are correctly added at level 0.

Fixes #4767.

### CODE REVIEW

**Comment by ajreynol** [comment]
The only thing that is suboptimal in this PR is how we now require 2 symbol tables in TPTP, due to how symbols are implicitly declared as they are parsed.

With further refactoring this could be merged into one, although I'm not sure if its very high priority.

File: src/context/cdhashmap.h [hunk 1]
```diff
@@ -211,8 +211,7 @@ class CDOhash_map : public ContextObj
     {
       // "Initializing" map insertion: this entry will never be
       // removed from the map, it's inserted at level 0 as an
-      // "initializing" element.  See
-      // CDHashMap<>::insertAtContextLevelZero().
+      // "initializing" element.
```
Comment (4tXJ7f) [inline src/context/cdhashmap.h:214]: From a cursory look, it seems that we might be able to remove the `atLevelZero` argument. Could you give that a try?

File: src/context/cdinsert_hashmap.h [hunk 2]
```diff
@@ -191,22 +190,13 @@ class CDInsertHashMap : public ContextObj {
...
-   * d_size and d_pushFronts are needed in restore.
+   * d_size are needed in restore.
    */
-  CDInsertHashMap(const CDInsertHashMap& l) :
-    ContextObj(l),
-    d_insertMap(NULL),
-    d_size(l.d_size),
-    d_pushFronts(l.d_pushFronts)
+  CDInsertHashMap(const CDInsertHashMap& l)
+      : ContextObj(l), d_insertMap(NULL), d_size(l.d_size)
```
Comment (4tXJ7f) [inline src/context/cdinsert_hashmap.h:199]:
```suggestion
      : ContextObj(l), d_insertMap(nullptr), d_size(l.d_size)
```

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
Cool, just some small requests.

**Review by 4tXJ7f — APPROVED** [review]
LGTM, thanks for the updates.

**Pull request merged**